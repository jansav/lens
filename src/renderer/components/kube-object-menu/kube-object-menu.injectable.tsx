/**
 * Copyright (c) 2021 OpenLens Authors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
import React from "react";

import {
  KubeObjectMenu,
  KubeObjectMenuDependencies,
  KubeObjectMenuProps,
} from "./kube-object-menu";

import type { KubeObject } from "../../../common/k8s-api/kube-object";
import { lifecycleEnum, Injectable } from "@ogre-tools/injectable";
import clusterNameInjectable from "./dependencies/cluster-name.injectable";
import kubeObjectMenuItemsInjectable from "./dependencies/kube-object-menu-items/kube-object-menu-items.injectable";
import removeActionInjectable from "./dependencies/remove-action/remove-action.injectable";
import updateActionInjectable from "./dependencies/update-action/update-action.injectable";

const KubeObjectMenuInjectable: Injectable<
  JSX.Element,
  KubeObjectMenuDependencies,
  KubeObjectMenuProps<KubeObject>
> = {
  getDependencies: (di, props) => ({
    clusterName: di.inject(clusterNameInjectable),

    removeObject: di.inject(removeActionInjectable, {
      customAction: props.removeAction,
      kubeObject: props.object,
    }),

    updateObject: di.inject(updateActionInjectable, {
      kubeObject: props.object,
    }),

    kubeObjectMenuItems: di.inject(kubeObjectMenuItemsInjectable, {
      kubeObject: props.object,
    }),
  }),

  instantiate: (dependencies, props) => (
    <KubeObjectMenu {...dependencies} {...props} />
  ),

  lifecycle: lifecycleEnum.transient,
};

export default KubeObjectMenuInjectable;
