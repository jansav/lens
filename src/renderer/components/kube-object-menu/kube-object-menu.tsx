/**
 * Copyright (c) 2021 OpenLens Authors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { boundMethod, cssNames } from "../../utils";
import type { KubeObject } from "../../../common/k8s-api/kube-object";
import { MenuActions, MenuActionsProps } from "../menu/menu-actions";
import identity from "lodash/identity";

import type { KubeObjectMenuRegistry } from "../../../extensions/registries";
import type { CatalogEntityRegistry } from "../../api/catalog-entity-registry";
import type { IGettableStore } from "../../../common/k8s-api/api-manager";


export interface KubeObjectMenuProps<T> extends MenuActionsProps {
  object: T | null | undefined;
  editable?: boolean;
  removable?: boolean;
}

interface KubeObjectMenuDependencies<T> extends KubeObjectMenuProps<T>{
  apiManager: IGettableStore,
  hideDetails: Function,
  editResourceTab: Function,
  catalogEntityRegistry: CatalogEntityRegistry,
  kubeObjectMenuRegistry: KubeObjectMenuRegistry
}

export class KubeObjectMenu<T extends KubeObject> extends React.Component<KubeObjectMenuDependencies<T>> {
  get dependencies() {
    const {
      apiManager,
      hideDetails,
      editResourceTab,
      catalogEntityRegistry,
      kubeObjectMenuRegistry,
    } = this.props;

    return {
      apiManager,
      editResourceTab,
      hideDetails,
      kubeObjectMenuRegistry,
      catalogEntityRegistry,
    };
  }

  get store() {
    const { object } = this.props;

    if (!object) return null;

    return this.dependencies.apiManager.getStore(object.selfLink);
  }

  get isEditable() {
    return this.props.editable ?? Boolean(this.store?.patch);
  }

  get isRemovable() {
    return this.props.removable ?? Boolean(this.store?.remove);
  }

  @boundMethod
  async update() {
    this.dependencies.hideDetails();
    this.dependencies.editResourceTab(this.props.object);
  }

  @boundMethod
  async remove() {
    this.dependencies.hideDetails();
    const { object, removeAction } = this.props;

    if (removeAction) await removeAction();
    else await this.store.remove(object);
  }

  @boundMethod
  getRemoveMessage() {
    const { object } = this.props;

    if (!object) {
      return null;
    }

    const breadcrumbParts = [
      this.dependencies.catalogEntityRegistry.activeEntity?.metadata?.name,
      object.getNs(),
      object.kind,
      object.getName(),
    ];

    const breadcrumb = breadcrumbParts.filter(identity).join("/");

    return (
      <p>Remove <b>{breadcrumb}</b>?</p>
    );
  }

  getMenuItems(): React.ReactChild[] {
    const { object, toolbar } = this.props;

    if (!object) {
      return [];
    }

    return this.dependencies.kubeObjectMenuRegistry
      .getItemsForKind(object.kind, object.apiVersion)
      .map(({ components: { MenuItem }}, index) => (
        <MenuItem
          object={object}
          key={`menu-item-${index}`}
          toolbar={toolbar}
        />
      ));
  }

  render() {
    const { remove, update, getRemoveMessage, isEditable, isRemovable } = this;
    const { className, editable, removable, ...menuProps } = this.props;

    return (
      <MenuActions
        className={cssNames("KubeObjectMenu", className)}
        updateAction={isEditable ? update : undefined}
        removeAction={isRemovable ? remove : undefined}
        removeConfirmationMessage={getRemoveMessage}
        {...menuProps}
      >
        {this.getMenuItems()}
      </MenuActions>
    );
  }
}
