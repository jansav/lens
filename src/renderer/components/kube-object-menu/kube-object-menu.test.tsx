/**
 * Copyright (c) 2021 OpenLens Authors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { KubeObjectMenu } from "./kube-object-menu";
import { KubeObject } from "../../../common/k8s-api/kube-object";
import userEvent from "@testing-library/user-event";
import type { KubeApi } from "../../../common/k8s-api/kube-api";
import type {
  IHasGettableItemsForKind,
  KubeObjectMenuRegistration,
} from "../../../extensions/registries";
import type { IGettableStore } from "../../../common/k8s-api/api-manager";
import type { KubeObjectStore } from "../../../common/k8s-api/kube-object.store";
import type { IHasName } from "../../../main/cluster";
import { ConfirmDialog } from "../confirm-dialog";
import asyncFn from "@async-fn/jest";

describe("kube-object-menu", () => {
  let hideDetailsStub: () => void;
  let editResourceTabStub: () => void;
  let apiManagerStub: IGettableStore;
  let clusterStub: IHasName;
  let kubeObjectMenuRegistryStub: IHasGettableItemsForKind;
  let objectStub: KubeObject | null;

  beforeEach(() => {
    // TODO: Remove illegal global overwrites for what should be a dependency somewhere.
    // TODO: Remove usage of experimental browser API.
    window.requestIdleCallback = (callback: IdleRequestCallback): number => {
      callback(undefined);

      return undefined;
    };

    window.cancelIdleCallback = () => {};

    apiManagerStub = {
      // eslint-disable-next-line unused-imports/no-unused-vars-ts

      getStore: <TKubeObjectStore extends KubeObjectStore<KubeObject>>(
        api: string | KubeApi<KubeObject>,
      ): TKubeObjectStore | undefined => undefined,
    };

    clusterStub = { name: "Some cluster name" };

    const MenuItemComponentStub: React.FC = () => <div>Some menu item</div>;

    const dynamicMenuItemStub: KubeObjectMenuRegistration = {
      apiVersions: ["irrelevant"],
      components: { MenuItem: MenuItemComponentStub },
      kind: "irrelevant",
    };

    kubeObjectMenuRegistryStub = {
      // eslint-disable-next-line unused-imports/no-unused-vars-ts
      getItemsForKind: (kind: string, apiVersion: string): any => [
        dynamicMenuItemStub,
      ],
    };

    hideDetailsStub = () => {};

    editResourceTabStub = () => {};
  });

  it("given no kube object, renders", () => {
    objectStub = null;

    const { baseElement } = render(
      <KubeObjectMenu
        object={objectStub}
        apiManager={apiManagerStub}
        cluster={clusterStub}
        kubeObjectMenuRegistry={kubeObjectMenuRegistryStub}
        hideDetails={hideDetailsStub}
        editResourceTab={editResourceTabStub}
        toolbar={true}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  describe("given kube object", () => {
    let baseElement: Element;
    let removeActionMock: any;
    let getByTestId: (arg0: string) => any;
    let queryByTestId: (arg0: string) => any;

    beforeEach(async () => {
      objectStub = KubeObject.create({
        apiVersion: "some-api-version",
        kind: "some-kind",
        metadata: {
          uid: "some-uid",
          name: "some-name",
          resourceVersion: "some-resource-version",
        },
      });

      removeActionMock = asyncFn();

      ({ baseElement, getByTestId, queryByTestId } = render(
        <div>
          <ConfirmDialog/>

          <KubeObjectMenu
            object={objectStub}
            apiManager={apiManagerStub}
            cluster={clusterStub}
            kubeObjectMenuRegistry={kubeObjectMenuRegistryStub}
            hideDetails={hideDetailsStub}
            editResourceTab={editResourceTabStub}
            toolbar={true}
            removeAction={removeActionMock}
          />
        </div>,
      ));
    });

    it("renders", () => {
      expect(baseElement).toMatchSnapshot();
    });

    it("does not open a confirmation dialog yet", () => {
      expect(queryByTestId("confirmation-dialog")).toBeNull();
    });

    describe("when removing kube object", () => {
      beforeEach(() => {
        const menuItem = getByTestId("menu-action-remove");

        userEvent.click(menuItem);
      });

      it("renders", () => {
        expect(baseElement).toMatchSnapshot();
      });

      it("opens a confirmation dialog", () => {
        getByTestId("confirmation-dialog");
      });

      describe("when remove is confirmed", () => {
        beforeEach(() => {
          const confirmRemovalButton = getByTestId("confirm");

          userEvent.click(confirmRemovalButton);
        });

        it("calls for removal of the kube object", () => {
          expect(removeActionMock).toHaveBeenCalledWith();
        });

        it("does not close the confirmation dialog yet", () => {
          getByTestId("confirmation-dialog");
        });

        it("when removal resolves, closes the confirmation dialog", async () => {
          await removeActionMock.resolve();

          expect(queryByTestId("confirmation-dialog")).toBeNull();        });
      });
    });
  });
});
