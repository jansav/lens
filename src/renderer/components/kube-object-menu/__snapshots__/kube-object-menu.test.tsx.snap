// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kube-object-menu given kube object renders 1`] = `
<body>
  <div>
    <div>
      <ul
        class="Animate opacity Menu MenuActions flex KubeObjectMenu toolbar gaps right bottom"
      >
        <div>
          Some menu item
        </div>
        <li
          class="MenuItem"
          data-menu-action-remove-test="true"
          data-testid="menu-action-remove"
          tabindex="0"
        >
          <i
            class="Icon material interactive focusable"
            id="tooltip_target_3"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="delete"
            >
              delete
            </span>
            <div />
          </i>
          <span
            class="title"
          >
            Remove
          </span>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="Animate opacity-scale Dialog flex center ConfirmDialog modal"
  />
  <div
    class="Tooltip narrow invisible formatter"
  >
    Delete
  </div>
</body>
`;

exports[`kube-object-menu given kube object when removing kube object renders 1`] = `
<body>
  <div>
    <div>
      <ul
        class="Animate opacity Menu MenuActions flex KubeObjectMenu toolbar gaps right bottom"
      >
        <div>
          Some menu item
        </div>
        <li
          class="MenuItem"
          data-menu-action-remove-test="true"
          data-testid="menu-action-remove"
          tabindex="0"
        >
          <i
            class="Icon material interactive focusable"
            id="tooltip_target_7"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="delete"
            >
              delete
            </span>
            <div />
          </i>
          <span
            class="title"
          >
            Remove
          </span>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="Animate opacity-scale Dialog flex center ConfirmDialog modal"
    data-testid="confirmation-dialog"
  >
    <div
      class="box"
    >
      <div
        class="confirm-content"
      >
        <i
          class="Icon material focusable big"
        >
          <span
            class="icon"
            data-icon-name="warning"
          >
            warning
          </span>
        </i>
         
        <div>
          <p>
            Remove 
            some-kind
             
            <b>
              some-name
            </b>
             from 
            <b>
              Some cluster name
            </b>
            ?
          </p>
        </div>
      </div>
      <div
        class="confirm-buttons"
      >
        <button
          class="Button cancel plain"
          type="button"
        >
          Cancel
        </button>
        <button
          class="Button ok primary"
          data-testid="confirm"
          type="button"
        >
          Remove
        </button>
      </div>
    </div>
  </div>
  <div
    class="Tooltip narrow invisible formatter"
  >
    Delete
  </div>
</body>
`;

exports[`kube-object-menu given no kube object, renders 1`] = `
<body>
  <div>
    <ul
      class="Animate opacity Menu MenuActions flex KubeObjectMenu toolbar gaps right bottom"
    />
  </div>
</body>
`;
